fix(bug): Adding check for None column value

iProgramStuff opened the pull request:
I ran into a bug where this line of code was running but the value of the column was None.

Full disclosure: Not sure if I called the API incorrectly or not...but this fixed my issue and the results seem to be correct.

fpacifici commented:
Could you provide the query you ran please ?

iProgramStuff commented:
> Could you provide the query you ran please ?

Working on it - was already on it, but ran into issues locally when testing and reverting changes. Will get to you asap.

tkaemming commented:
Yeah, I'm a little confused about how this would happen as well — the only thing that I can think of is if something was aliased as an empty string or explicit null?

iProgramStuff commented:
```{
   "orderby":[
      "events.last_seen",
      "events.issue"
   ],
   "consistent":false,
   "having":[

   ],
   "aggregations":[
      [
         "multiply(toUInt64(max(events.timestamp)), 1000)",
         "",
         "events.last_seen"
      ],
      [
         "uniq",
         "events.issue",
         "events.total"
      ]
   ],
   "totals":true,
   "project":[
      2,
      3,
      4,
      5,
      6,
      7
   ],
   "from_date":"2019-10-07T21:04:00.121516",
   "selected_columns":[

   ],
   "limit":69,
   "to_date":"2019-10-21T21:04:00.121516",
   "offset":0,
   "conditions":[
      [
         "groups.status",
         "=",
         0
      ],
      [
         "environment",
         "IN",
         [
            "staging",
            "beta"
         ]
      ],
      [
         "project_id",
         "IN",
         [
            7,
            6,
            5,
            4,
            3,
            2
         ]
      ],
      [
         "events.issue",
         "IN",
         [
            1,
            2,
            3,
            4,
            5,
            6,
            7,
            8,
            10,
            11,
            12,
            13,
            14,
            15,
            16,
            17,
            18,
            19,
            20,
            21,
            22,
            23,
            24,
            25,
            26,
            27,
            28,
            29,
            30,
            31,
            32,
            33,
            34,
            35,
            36,
            37,
            38,
            39,
            40,
            41,
            42,
            43,
            44,
            45,
            46,
            47,
            48,
            49,
            54,
            58,
            59,
            60,
            61,
            62,
            85,
            86,
            89,
            90,
            92,
            93,
            120,
            121,
            122,
            123,
            128,
            154,
            156,
            159,
            160
         ]
      ]
   ],
   "groupby":[
      "events.issue",
      "groups.first_seen",
      "groups.status"
   ]
}```
is the query I can enter into `http://localhost:1218/groups/query` to reproduce this error with my database. Wondering if I have bad data in mine.

The row that gives an issue is:
{'events.issue': None, 'groups.first_seen': None, 'groups.status': None, 'events.last_seen': 0, 'events.total': 0}
The column is: `groups.first_seen`. The column type is `Nullable(DateTime)`. It passes the regex (DATETIME_TYPE_RE.match(col["type"]):) and then tries to call .replace on None.

I think it is because `totals: true`. as the clickhouse result for the query ```SELECT
    nullIf(events.group_id, 0) AS `events.issue`,
    groups.first_seen AS `groups.first_seen`,
    groups.status AS `groups.status`,
    toUInt64(max(events.timestamp)) * 1000 AS `events.last_seen`,
    ifNull(uniq(nullIf(events.group_id, 0)), 0) AS `events.total`
FROM groupedmessage_local AS groups
LEFT JOIN sentry_local AS events ON (groups.project_id = events.project_id) AND (groups.id = events.group_id)
WHERE (`groups.status` = 0) AND (environment IN ('staging', 'beta')) AND (project_id IN (7, 6, 5, 4, 3, 2)) AND (`events.issue` IN (1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 54, 58, 59, 60, 61, 62, 85, 86, 89, 90, 92, 93, 120, 121, 122, 123, 128, 154, 156, 159, 160)) AND (project_id IN (2, 3, 4, 5, 6, 7)) AND ((events.timestamp AS `events.timestamp`) >= toDateTime('2019-10-07T21:04:00')) AND (`events.timestamp` < toDateTime('2019-10-21T21:04:00')) AND ((events.deleted AS `events.deleted`) = 0) AND ((groups.record_deleted AS `groups.record_deleted`) = 0)
GROUP BY (`events.issue`, `groups.first_seen`, `groups.status`)
    WITH TOTALS
ORDER BY
    `events.last_seen` ASC,
    `events.issue` ASC
LIMIT 0, 69```
is:
![image](https://user-images.githubusercontent.com/6395250/67243707-36647a80-f426-11e9-9a49-0aae5fd9008b.png)
(where `groups.last_seen` is null)

tkaemming commented:
I misread this as checking that the column _name_ was none — ignore my previous comment. 😬

tkaemming commented:
I think this is generally fine, but we should only be applying this if the column type itself is nullable in my opinion.

iProgramStuff commented:
> 
> 
> I think this is generally fine, but we should only be applying this if the column type itself is nullable in my opinion.

I think the field in totals can be null even if the column itself cannot be null.

fpacifici commented on snuba/reader.py:
```diff
@@ -70,9 +70,10 @@ def iterate_rows():
+                if row[col["name"]] is not None:  # The column value can be null/None at times.
+                    row[col["name"]] = (
+                        row[col["name"]].replace(tzinfo=tz.tzutc()).isoformat()
+                    )
```
Ok, good catch, I think we did not experience this before because we had only one nullable datetime column before CDC and that was pretty much never null, also likely we were not grouping by a datetime column with totals (the other way you can get None here).

Before  going on, 
- let's do as Ted said and skip the type conversion only if the column is actually Nullable. It is easy since the regex above is already considering that. The result of the match already contains a `Nullable(` group if the column is nullable
- please do the same for DATE_TYPE_RE (right below)
- please let's cover this case in the unit test: tests/test_api.py::TestApi::test_totals. It should be a trivial addition. if you group by a datetime column you should be stressing this.

thanks

fpacifici commented:
About the nullability of the column. We are actually looking at the data type in the result:
```
{
    'data': [
           ....
     ],
     'meta': [
        {'name': 'events.issue', 'type': 'Nullable(UInt64)'}, 
        {'name': 'groups.first_seen', 'type': 'Nullable(DateTime)'}, 
        {'name': 'groups.status', 'type': 'Nullable(UInt8)'}, 
        {'name': 'events.last_seen', 'type': 'UInt64'}, 
        {'name': 'events.total', 'type': 'UInt64'}
    ], 
    'totals': {
        'events.issue': None, 
        'groups.first_seen': None, 
        'groups.status': None, 
        'events.last_seen': 1571288640000, 
        'events.total': 15
    }
}
```

`{'name': 'events.issue', 'type': 'Nullable(UInt64)'}, ` is not a nullable column in the dataset but it is in the result of this query (because of the totals clause). The type to look for is the one inferred by clickhouse in the "meta" section of the response.

fpacifici commented on snuba/reader.py:
```diff
@@ -70,16 +70,18 @@ def iterate_rows():
-                row[col["name"]] = (
-                    row[col["name"]].replace(tzinfo=tz.tzutc()).isoformat()
-                )
+                if row[col["name"]] is not None:  # The column value can be null/None at times.
```
This is a valid condition only if the regex match above contains "Nullable" otherwise, this being null, would be an error. I think we should add that check
